File: leetcode/binary_search/2137_Pour_Water_Between_Buckets_to_Make_Water_Levels_Equal/codes/bs.c
#include <stdio.h>
#include <math.h>

double equalizeWater(int* buckets, int bucketsSize, int loss) {
    double lo = 0.0, hi = 0.0;
    for (int i = 0; i < bucketsSize; i++)
        if (buckets[i] > hi) hi = buckets[i];
    double eps = 1e-6;
    double factor = (100.0 - loss) / 100.0;

    while (hi - lo > eps) {
        double mid = (lo + hi) / 2.0;
        double need = 0.0, have = 0.0;
        for (int i = 0; i < bucketsSize; i++) {
            if (buckets[i] < mid) need += mid - buckets[i];
            else have += buckets[i] - mid;
        }
        if (have * factor >= need)
            lo = mid;
        else
            hi = mid;
    }
    return lo;
}

File: leetcode/binary_search/2137_Pour_Water_Between_Buckets_to_Make_Water_Levels_Equal/codes/bs.py
from typing import List

class Solution:
    def equalizeWater(self, buckets: List[int], loss: int) -> float:
        lo, hi = 0.0, max(buckets)
        eps = 1e-6
        factor = (100 - loss) / 100.0

        def can(mid):
            need = 0.0
            have = 0.0
            for b in buckets:
                if b < mid:
                    need += mid - b
                else:
                    have += b - mid
            return have * factor >= need

        while hi - lo > eps:
            mid = (lo + hi) / 2
            if can(mid):
                lo = mid
            else:
                hi = mid
        return lo
